File: apps/api/src/InventoryPro.Domain/Catalog/Product.cs
namespace InventoryPro.Domain.Catalog;

using InventoryPro.Domain.Common;

/// <summary>
/// Phân loại sản phẩm.
/// </summary>
public enum ProductType
{
    Goods = 0,
    Service = 1,
    RawMaterial = 2,
    FinishedGood = 3,
    Consumable = 4,
}

/// <summary>
/// Trạng thái sản phẩm.
/// </summary>
public enum ProductStatus
{
    Active = 0,
    Inactive = 1,
    Archived = 2,
}

/// <summary>
/// Sản phẩm / vật tư. SKU unique theo tenant.
/// </summary>
public class Product : TenantScopedEntity
{
    public string Sku { get; set; } = string.Empty;
    public string? Barcode { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    public Guid? CategoryId { get; set; }
    public Guid BaseUnitId { get; set; }
    public ProductType ProductType { get; set; } = ProductType.Goods;
    public decimal CostPrice { get; set; } = 0;
    public decimal SellPrice { get; set; } = 0;
    public decimal MinStock { get; set; } = 0;
    public decimal? MaxStock { get; set; }
    public bool IsBatchTracked { get; set; } = false;
    public bool IsSerialTracked { get; set; } = false;
    public bool IsExpiryTracked { get; set; } = false;
    public decimal? Weight { get; set; }      // gram
    public decimal? Volume { get; set; }      // cm3
    public string Attributes { get; set; } = "{}";  // JSONB string
    public string? ImageUrl { get; set; }
    public ProductStatus Status { get; set; } = ProductStatus.Active;

    // Navigation
    public Category? Category { get; set; }
    public UnitOfMeasure? BaseUnit { get; set; }
    public ICollection<ProductUnit> Units { get; set; } = new List<ProductUnit>();
}

File: apps/api-worker/src/validators/product.ts
/**
 * Zod validation schemas cho Product module
 *
 * - listProductsQuery: validate query params (page, pageSize, search, ...)
 * - createProductRequest: validate body cho POST /products
 * - updateProductRequest: validate body cho PUT /products/:id
 *
 * Match shape với frontend types (apps/web/src/features/products/api.ts)
 */

import { z } from "zod";

const uuid = z.string().uuid();
const decimalString = z
  .union([z.string(), z.number()])
  .transform((v) => String(v));

export const listProductsQuery = z.object({
  page: z.coerce.number().int().min(1).default(1),
  pageSize: z.coerce.number().int().min(1).max(100).default(20),
  search: z.string().trim().optional(),
  categoryId: uuid.optional(),
  status: z.enum(["ACTIVE", "INACTIVE", "ARCHIVED"]).optional(),
  productGroup: z
    .enum(["HOA_CHAT_SINH_PHAM", "VAT_TU_Y_TE", ""])
    .optional()
    .transform((v) => (v === "" ? undefined : v)),
  isActive: z.coerce.boolean().optional(),
});

export const createProductRequest = z.object({
  sku: z.string().trim().min(1).max(50),
  barcode: z.string().trim().max(50).optional().nullable(),
  name: z.string().trim().min(1).max(200),
  description: z.string().trim().max(2000).optional().nullable(),
  categoryId: uuid.optional().nullable(),
  baseUnitId: uuid,
  productType: z
    .enum(["GOODS", "SERVICE", "RAW_MATERIAL", "FINISHED_GOOD", "CONSUMABLE"])
    .default("GOODS"),
  costPrice: decimalString.default("0"),
  sellPrice: decimalString.default("0"),
  minStock: decimalString.default("0"),
  maxStock: decimalString.optional().nullable(),
  isBatchTracked: z.boolean().default(false),
  isSerialTracked: z.boolean().default(false),
  isExpiryTracked: z.boolean().default(false),
  status: z.enum(["ACTIVE", "INACTIVE", "ARCHIVED"]).default("ACTIVE"),
  imageUrl: z.string().url().optional().nullable(),
  // Khoa XN
  productGroup: z
    .enum(["HOA_CHAT_SINH_PHAM", "VAT_TU_Y_TE"])
    .optional()
    .nullable(),
  productSubtype: z.string().trim().max(50).optional().nullable(),
  openVialStabilityDays: z.coerce.number().int().min(0).max(365).optional().nullable(),
  storageCondition: z
    .enum(["ROOM_TEMP", "REFRIGERATED", "FROZEN"])
    .optional()
    .nullable(),
});

export const updateProductRequest = createProductRequest.partial();

export type ListProductsQuery = z.infer<typeof listProductsQuery>;
export type CreateProductRequest = z.infer<typeof createProductRequest>;
export type UpdateProductRequest = z.infer<typeof updateProductRequest>;
